sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -1737,3 +1738,33 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
     }
   }
 }
+
+class ThreadLeakInSortMergeJoinSuite
+  extends QueryTest
+    with SharedSparkSession
+    with AdaptiveSparkPlanHelper {
+
+  setupTestData()
+  override protected def createSparkSession: TestSparkSession = {
+    SparkSession.cleanupAnyExistingSession()
+    new TestSparkSession(
+      sparkConf.set(SHUFFLE_SPILL_NUM_ELEMENTS_FORCE_SPILL_THRESHOLD, 20))
+  }
+
+  test("SPARK-47146: thread leak when doing SortMergeJoin (with spill)") {
+
+    withSQLConf(
+      SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "1") {
+
+      assertSpilled(sparkContext, "inner join") {
+        sql("SELECT * FROM testData JOIN testData2 ON key = a").collect()
```

Comment:
This test will work in current master as well, right ? Since it reads the iterator entirely and so will get closed ?
The test should be failing in master and work with this fix @JacobZheng0927

Reply:
No, this test fails in master because it doesn't read the iterator completely so there is a thread leak

Reply:
Can you explain why it wont read the iterator fully ? The `collect` should drop it entirely ?

Reply:
As an example, sort merge join performs an inner join with the keys [1, 2, 3, 4, 5, 6] and [1, 2, 3] on the left and right. The left and right iterators will take turns advancing. The right iterator will hit the end first, and there's no point in advancing the left iterator.

Reply:
Sounds good, does that happen here though ? What I am trying to make sure is - does this test actually fail in master and work after this PR ? (not at desktop, so cant run the test against master to check :-) )

Reply:
Yes, this test will fail in master and pass after this PR.